source/slang/ir.cpp

```diff
@@ -3103,10 +3103,30 @@ namespace Slang
     IRGlobalVar* cloneGlobalVar(IRSpecContext* context, IRGlobalVar* originalVar);
     IRFunc* cloneFunc(IRSpecContext* context, IRFunc* originalFunc);
     IRWitnessTable* cloneWitnessTable(IRSpecContext* context, IRWitnessTable* originalVar);
+    RefPtr<Substitutions> cloneSubstitutions(
+        IRSpecContext*  context,
+        Substitutions*  subst);
 
     RefPtr<Type> IRSpecContext::maybeCloneType(Type* originalType)
     {
-        return originalType->Substitute(subst).As<Type>();
+        auto rsType = originalType->GetCanonicalType()->Substitute(subst).As<Type>();
+        if (auto declRefType = rsType.As<DeclRefType>())
+        {
+            if (subst)
+            {
+                auto newSubst = cloneSubstitutions(this, subst);
+                insertSubstAtBottom(declRefType->declRef.substitutions, newSubst);
+            }
+        }
+        else if (auto funcType = rsType.As<FuncType>())
```

Comment:
This code seems to be doing an ad hoc structural recursion on types (relying on the fact that 99% of types are `DeclRefType`s and that `FuncType` is the main structural type).

It really seems to be that the right answer here is not to be doing this recursive step here, but instead to fold the required logic into the existing `Substitute`. Basically, it seems like the `insertSubstAtBottom` sort of logic should be the default behavior for a `DeclRefType` when encountering any `Substitutions` list that includes global generic parameter substitutions.

I'm fine with us letting this in as-is, but we really need to start working on being less ad hoc about these things and trying to make the existing `DeclRef`/`Substitutions` machinery do the Right Thing instead. Otherwise we are going to find ourselves having to make similar special-case modifications all over the place (as is currently done for `ThisTypeSubstitution`s).

Reply:
I think what makes sense here is to make FuncType a DeclRefType so we can store the substitutions in it.

Reply:
`FuncType::SubstituteImpl()` already recursively applies substitution to both the parameter and result types, to create a new `FuncType`. That means it will automatically do the right thing as long as the leaf cases do the right thing.

I think that the right place to handle this is as part of `Substitutions::SubstituteImpl()`. Right now that is a `virtual` function that includes a lot of duplication across the different sub-classes, but it seems like we should bottleneck it through a common routine.

The root problem, as I understand it, is lines like the following one in `GenericSubstitution::SubstituteImpl()`:

```c++
auto outerSubst = outer ? outer->SubstituteImpl(subst, &diff) : nullptr;
```

In the case where `outer` is null, this should *not* just be yielding an empty/null substitution, but rather a substitution that includes only the `GlobalGenericParameterSubstitution`s from `subst` (these should always be the outer-most substitutions anyway, so we can probably just re-use them by pointer).

That basically tells you the general case of what to do when applying substitutions to existing substitutions:

- If the target substitution list is null/empty, then the result should be the global-type-parameter substitutions from the source substitution list

- Otherwise, apply the source substitutions to the `outer` substitution of the target (if `outer` is null then the above case will apply to it)

- Now apply the subclass-specific substitution logic to the target, passing in both the source substitution and the already-substituted `outer`

That approach should eliminate some of the duplication between the cases, and also wrap up global-type-parameter substitution as part of the ordinary logic. Right?

Reply:
Yeah, that sounds right. I will try doing it this way and see whether it triggers any other issues.